src/Compilers/CSharp/Portable/Symbols/FieldSymbol.cs

```diff
@@ -464,15 +466,34 @@ public virtual int TupleElementIndex
             get
             {
                 // wrapped tuple fields already have this information and override this property
-                Debug.Assert(!(this is TupleElementFieldSymbol or TupleErrorFieldSymbol));
+                Debug.Assert(!(this is TupleElementFieldSymbol or TupleErrorFieldSymbol or Retargeting.RetargetingFieldSymbol));
+                if (!ContainingType.IsTupleType)
+                {
+                    return -1;
+                }
 
-                var map = ContainingType.TupleFieldDefinitionsToIndexMap;
-                if (map is object && map.TryGetValue(this.OriginalDefinition, out int index))
+                if (!ContainingType.IsDefinition)
                 {
-                    return index;
+                    return this.OriginalDefinition.TupleElementIndex;
                 }
 
-                return -1;
+                var tupleElementPosition = NamedTypeSymbol.MatchesCanonicalTupleElementName(Name);
+                int arity = ContainingType.Arity;
+                if (tupleElementPosition <= 0 || tupleElementPosition > arity)
+                {
+                    // ex: no "Item2" in 'ValueTuple<T1>'
+                    return -1;
+                }
+                Debug.Assert(tupleElementPosition < NamedTypeSymbol.ValueTupleRestPosition);
```

Comment:
Could you please clarify why we can assume this?

Reply:
Should this assert be moved to immediately following the `if (!ContainingType.IsDefinition) { ... }` block?